kpi/deployment_backends/kobocat_backend.py

```diff
@@ -724,6 +725,100 @@ def set_validation_statuses(self, data, user, method):
         kc_response = self.__kobocat_proxy_request(kc_request, user)
         return self.__prepare_as_drf_response_signature(kc_response)
 
+    def set_bulk_update_submissions(
+        self, request_data: dict, requesting_user_id: int
+    ) -> dict:
+        """
+        Allows for bulk updating of submissions proxied through kobocat. A
+        `deprecatedID` for each submission is given the previous value of
+        `instanceID` and `instanceID` receives an updated uuid. For each key
+        and value within `request_data`, either a new element is created on the
+        submission's XML tree, or the existing value is replaced by the updated
+        value.
+
+        Args:
+            request_data (dict): must contain a list of `submission_ids` and at
+                least one other key:value field for updating the submissions
+            requesting_user_id (int)
+
+        Returns:
+            dict: formatted dict to be passed to a Response object
+        """
+        payload = self.__prepare_payload(request_data)
+        kwargs = {'instance_ids': payload.pop('submission_ids')}
+        params = self.validate_submission_list_params(
+            requesting_user_id, format_type=INSTANCE_FORMAT_TYPE_XML, **kwargs
+        )
+        submissions = list(self.__get_submissions_in_xml(**params))
+        validated_submissions = self.__validate_bulk_update_submissions(
+            submissions
+        )
+
+        response_codes = []
+        for submission in validated_submissions:
+            xml_parsed = ET.fromstring(submission)
+
+            _uuid, uuid_formatted = self.generate_new_instance_id()
+
+            # updating xml fields for submission. In order to update an existing
+            # submission, the current `instanceID` must be moved to the value
+            # for `deprecatedID`.
+            instance_id = xml_parsed.find('./meta/instanceID')
+            # if the submission has been edited before, it will already contain
+            # a deprecatedID element - otherwise create a new element
+            deprecated_id = (
+                xml_parsed.find('./meta/deprecatedID')
+                if xml_parsed.find('./meta/deprecatedID') is not None
+                else ET.SubElement(xml_parsed.find('meta'), 'deprecatedID')
+            )
+            deprecated_id.text = instance_id.text
+            instance_id.text = uuid_formatted
+
+            # if the form has been updated with new fields and earlier
+            # submissions have been selected as part of the bulk update,
+            # a new element has to be created before a value can be set.
+            # However, with this new power, arbitrary fields can be added
+            # to the XML tree through the api.
```

Comment:
right, it's powerful, but that's what we want

Reply:
🤘